src/git.ts

```diff
@@ -0,0 +1,259 @@
+  private _cache: {repos: Map<string, Map<string, GitHub.Repo>>};
+  private _cloneOptions: nodegit.CloneOptions;
+  private _cloneRateLimiter: Bottleneck;
+  private _github: GitHub;
+  private _token: string;
+  private _user: GitHub.User;
+
+  constructor(token: string) {
+    this.resetCache();
+    this._token = token;
+    this._github = new GitHub({
+      version: '3.0.0',
+      protocol: 'https',
+    });
+    this._github.authenticate({type: 'oauth', token: token});
+    // TODO: Make the arguments to rate limiter configurable.
+    this._cloneRateLimiter = new Bottleneck(20, 100);
+    this._cloneOptions = {
+      fetchOpts: {
+        callbacks: {
+          certificateCheck: function() {
+            return 1;
+          },
+          credentials: function(url: string, userName: string) {
+            return nodegit.Cred.userpassPlaintextNew(token, 'x-oauth-basic');
+          }
+        }
+      }
+    };
+  }
+
+  resetCache() {
+    this._cache = {repos: new Map()};
+  }
+
+  // HACK: Don't expose repos, create the necessary methods to operate on them.
+  get repos() {
+    return this._github.repos;
+  }
+
+  // HACK: Don't expose user, create the necessary methods to operate on them.
+  get user() {
+    return this._github.user;
+  }
+
+  /**
+   * Given a github repository and a directory to clone it into, return an
+   * ElementRepo once it has been cloned and checked out.
+   */
+  async clone(githubRepo: GitHub.Repo, cloneDir: string):
+      Promise<nodegit.Repository> {
+    let nodegitRepo: nodegit.Repository;
+    if (util.existsSync(cloneDir)) {
+      let updatedRepo: nodegit.Repository;
+      nodegitRepo =
+          await nodegit.Repository.open(cloneDir)
+              .then((repo) => {
+                updatedRepo = repo;
+                return this._cloneRateLimiter.schedule(
+                    () => updatedRepo.fetchAll(this._cloneOptions.fetchOpts));
+              })
+              .then(() => updatedRepo);
+    } else {
+      // Potential race condition if multiple repos w/ the same name are
+      // checked
+      // out simultaneously.
+      nodegitRepo = await this._cloneRateLimiter.schedule(() => {
+        return nodegit.Clone.clone(
+            githubRepo.clone_url, cloneDir, this._cloneOptions);
+      });
+    }
+    return nodegitRepo;
+  }
+
+  /**
+   * @returns a representation of a github repo from a string version
+   */
+  async getRepoInfo(owner: string, repo: string): Promise<GitHub.Repo> {
+    if (this._cache.repos.has(owner.toLowerCase()) &&
+        this._cache.repos.get(owner.toLowerCase()).has(repo.toLowerCase())) {
+      return this._cache.repos.get(owner.toLowerCase()).get(repo.toLowerCase());
+    }
+    return promisify(this._github.repos.get)({user: owner, repo: repo})
+        .then((response) => {
+          // TODO(usergenic): Patch to _handle_ redirects and/or include
+          // details in error messaging.  This was encountered because we
```

Comment:
Not *hard*, but it would have to be extracted from the redirection url, since it's not given as a specific value in the response IIRC.